design-documents/distributed-deployment.md

```diff
@@ -0,0 +1,217 @@
+    
+    CatalogExtensionAdminUi/composer.json
+    ```
+    {
+        "name": "vendor/catalog-extension-admin-ui",
+        ...
+        "instances": [
+            "admin-ui"
+        ]
+    }
+    ```
+    
+    As when we install extension we use metapackage name (metapackage includes all parts of the extension: AdminUi, Ui, Cron, WebApi) and metapackage is not being installed or saved anywhere, we need to add this information to the composer.json on the instance.
+    
+    admin-ui-instance/composer.json
+    ```
+    {
+        "name": "magento/admin-ui-project",
+        "type": "project",
+        "require": {
+            "vendor/catalog-extension-admin-ui": "*"
+        },
+        "installed-packages": [
+            "vendor/catalog-extension"
+        ]
+    }
+    ```
+    
+    This would allow to understand which packages that are installed on the instance are part of extension we want to remove and remove these packages from all instances.
+    
+    Pros
+     * The same mechanism can be used for distributed deployment of services
+    
+    Cons
+    * Will require us to publish more projects
+    * Need to add new type of configuration
+    
+    Examples of using command
+    
+    ```
+    instance-manager install --package="vendor-name/catalog-customization" --instances "./admin-ui ./ui ./cron ./webapi"
+    instance-manager uninstall --package="vendor-name/catalog-customization" --instances "./admin-ui ./ui ./cron ./webapi"
+    instance-manager update --instances "./admin-ui ./ui ./cron ./webapi"
+    ```
+    
+    Some of corner cases tool need to support
+    1. When upgrading extension that shares common package with another extension, make sure that version of the common package satisfies metapackage of another package
+    2. When removing package from one of the instances that is part of the extension user want to remove, check that package is not part of any other installed extension
+    3. When updating packages, check that new versions of the packages satisfy version requirements of metapackage
+    4. When updating extension find new versions of extensions that consist from packages that have their dependency/versions met (dependencies on Magento packages). When calculating versions, we probably will have to take into account Magento package versions on other instances
+    
+    #### Tool architecture
+    There are two options for tool implementation
+    
+    1. Tool will be responsible for version calculation and constraint checks.
+    2. To perform operation, tool will be performing this operation on monoliths first and then updating instances using packages on monoliths as a master. Because of the complexity described above this approach might be more reliable.
+
+3. Convention based
+
+    We can have convention similar to this
+    * Modules that end with *admin-ui go to admin instance
+    * Modules that end with *ui and not *admin-ui go to store front instance
```

Comment:
We probably will have modules with UI logic applicable for admin and storefront.